Expose `Response`'s `Upgrader` field

GNMoseke opened the pull request:
Allows user access to the `upgrader` field on `Response` object, along with refactoring to the `Upgrader` protocol and the conforming `WebSocketUpgrader` struct. This allows for custom user-defined dynamic upgrade behavior in a route handler. For example, a user can now do something like this, with both behaviors on the same route:

```swift
    app.get("example") { req async throws -> Response in
        // if it's a socket request, handle it differently.
        if req.headers.contains(name: .secWebSocketProtocol) {
            resp.upgrader = WebSocketUpgrader(
                maxFrameSize: .init(integerLiteral: 1 << 12),
                shouldUpgrade: { req.eventLoop.makeSucceededFuture([:]) },
                onUpgrade: { ws in
                    // Do custom upgrade callback behavior here...
                }
            )
        }
        return resp
    }
```

0xTim commented:
Overall I'm happy to accept this. My one concern is adding a public enum with a case - that really locks us in to that enum and it's almost impossible to justify adding a new case as that would break the public API. If you could change it to be a static function/property somehow I'd be much happier. I'll have a think about the best way to achieve this unless you think of something first!

GNMoseke commented:
@0xTim Sorry for such a delay on getting back to this, I think that concern is totally fair. I went ahead and split it out into an `Upgrader` protocol that provides generic upgrade behavior definition and refactored the websocket enum case into a conforming struct. This way we keep all the existing behavior, make it easily extendable in the future, and allow users to define their own custom `Upgrader`s should they choose. Let me know if you think this approach looks better.

0xTim commented:
@GNMoseke yep looks much better thanks! Could you fix the indentation? It looks like some of the changes are double indented making it difficult to see what's new and what's existing. Thanks!

GNMoseke commented:
Should be good to go now.

GNMoseke commented on Sources/Vapor/HTTP/Server/HTTPServerUpgradeHandler.swift:
```diff
@@ -43,20 +43,14 @@ final class HTTPServerUpgradeHandler: ChannelDuplexHandler, RemovableChannelHand
-                    }, upgradePipelineHandler: { channel, req in
-                        return WebSocket.server(on: channel, onUpgrade: onUpgrade)
-                    })
+            let res = self.unwrapOutboundIn(data)
+            
+            // check upgrade
+            switch self.upgradeState {
+            case .pending(let req, let buffer):
+                self.upgradeState = .upgraded
+                if res.status == .switchingProtocols, let upgrader = res.upgrader {
+                    let protocolUpgrader = upgrader.applyUpgrade(req: req, res: res)
```
Are we looking at the same version of the diff? The most recent one looks fine to me. What specifically looks off?

0xTim replied:
<img width="642" alt="Screenshot 2022-06-01 at 16 55 23" src="https://user-images.githubusercontent.com/9938337/171447665-f6acfb28-4914-46df-b5e4-fe160065b22f.png">
 I think so? See how line 46  has extra indentation - it hasn't actually changed but GH thinks it has, right?

0xTim replied:
Also https://github.com/GNMoseke/vapor/blob/feature/expose-upgrader/Sources/Vapor/HTTP/Server/HTTPServerUpgradeHandler.swift#L46

0xTim reviewed (approved):
Thanks! This will get bundled in with several PRs tomorrow
